File: sdks/ts/memory/src/ingest/hash.test.ts
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it } from 'vitest'
import {
  blake3Hasher,
  hashChunk,
  hashDocument,
  hashDocumentId,
  hashSlug,
  hashString,
  type Hasher,
} from './hash.js'

describe('hashDocument', () => {
  it('produces deterministic 64-char hex output', () => {
    const input = Buffer.from('the quick brown fox jumps over the lazy dog')
    const first = hashDocument(input)
    const second = hashDocument(input)
    expect(first).toBe(second)
    expect(first).toHaveLength(64)
    expect(first).toMatch(/^[0-9a-f]{64}$/)
  })

  it('produces different hashes for different inputs', () => {
    const a = hashDocument(Buffer.from('input alpha'))
    const b = hashDocument(Buffer.from('input beta'))
    expect(a).not.toBe(b)
  })

  it('handles empty input without throwing', () => {
    const hash = hashDocument(Buffer.alloc(0))
    expect(hash).toHaveLength(64)
    expect(hash).toMatch(/^[0-9a-f]{64}$/)
  })

  it('handles large inputs', () => {
    const largeBuffer = Buffer.alloc(100 * 1024, 0x61)
    const hash = hashDocument(largeBuffer)
    expect(hash).toHaveLength(64)
    expect(hash).toMatch(/^[0-9a-f]{64}$/)
  })
})

describe('hashChunk', () => {
  it('produces the same output as hashDocument for identical input', () => {
    const input = Buffer.from('chunk content for verification')
    expect(hashChunk(input)).toBe(hashDocument(input))
  })

  it('produces 64-char hex output', () => {
    const hash = hashChunk(Buffer.from('some chunk'))
    expect(hash).toHaveLength(64)
    expect(hash).toMatch(/^[0-9a-f]{64}$/)
  })
})

describe('hashString', () => {
  it('produces the same output as hashDocument with Buffer.from(s, utf8)', () => {
    const s = 'string input for verification'
    const fromString = hashString(s)
    const fromBuffer = hashDocument(Buffer.from(s, 'utf8'))
    expect(fromString).toBe(fromBuffer)
  })

  it('produces 64-char hex output', () => {
    const hash = hashString('hello world')
    expect(hash).toHaveLength(64)
    expect(hash).toMatch(/^[0-9a-f]{64}$/)
  })

  it('handles empty string', () => {
    const hash = hashString('')
    expect(hash).toHaveLength(64)
    expect(hash).toMatch(/^[0-9a-f]{64}$/)
  })
})

describe('hashSlug', () => {
  it('produces 12-char hex output', () => {
    const hash = hashSlug(Buffer.from('some content for slug'))
    expect(hash).toHaveLength(12)
    expect(hash).toMatch(/^[0-9a-f]{12}$/)
  })

  it('is a prefix of hashDocument', () => {
    const input = Buffer.from('consistency check')
    const full = hashDocument(input)
    const slug = hashSlug(input)
    expect(full.slice(0, 12)).toBe(slug)
  })

  it('handles empty input', () => {
    const hash = hashSlug(Buffer.alloc(0))
    expect(hash).toHaveLength(12)
    expect(hash).toMatch(/^[0-9a-f]{12}$/)
  })
})

describe('hashDocumentId', () => {
  it('produces 16-char hex output', () => {
    const id = hashDocumentId('brain-1', 'abc123')
    expect(id).toHaveLength(16)
    expect(id).toMatch(/^[0-9a-f]{16}$/)
  })

  it('is brain-scoped: different brainIds yield different IDs', () => {
    const contentHash = 'deadbeefcafebabe'
    const idA = hashDocumentId('brain-alpha', contentHash)
    const idB = hashDocumentId('brain-beta', contentHash)
    expect(idA).not.toBe(idB)
  })

  it('different brains with same content produce different IDs', () => {
    const contentHash = hashDocument(Buffer.from('shared document content'))
    const brains = ['brain-a', 'brain-b', 'brain-c', 'brain-d', 'brain-e']
    const ids = brains.map((b) => hashDocumentId(b, contentHash))
    const unique = new Set(ids)
    expect(unique.size).toBe(brains.length)
  })
})

describe('Hasher interface', () => {
  it('blake3Hasher satisfies Hasher type', () => {
    const h: Hasher = blake3Hasher
    expect(h.name).toBe('blake3')
  })

  it('blake3Hasher.hash matches hashDocument', () => {
    const input = Buffer.from('hasher interface test')
    expect(blake3Hasher.hash(input)).toBe(hashDocument(input))
  })
})

describe('cross-SDK conformance', () => {
  it('produces identical hash for the canonical test vector across Go and TS', () => {
    // This test vector MUST produce the same hash in both the Go and TS SDKs.
    // The Go test logs this value; if it ever changes, both tests must be
    // updated in lockstep.
    const input = "jeff's brain memory system"
    const hash = hashString(input)

    // BLAKE3("jeff's brain memory system") — verified against Go SDK output.
    // If this assertion fails after a dependency update, re-run the Go test
    // to obtain the correct expected value.
    expect(hash).toHaveLength(64)
    expect(hash).toMatch(/^[0-9a-f]{64}$/)

    // Hardcoded expected value from BLAKE3 reference implementation.
    // Computed via: blake3.Sum256([]byte("jeff's brain memory system"))
    expect(hash).toBe(
      'e311e54b56b26bfef4e5c8501f04c708f1e02233106022f58a7e94b728b7265c',
    )
  })
})

File: go/ingest/hash_test.go
// SPDX-License-Identifier: Apache-2.0
package ingest

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
	"testing"
)

func TestHashDocument_Deterministic(t *testing.T) {
	t.Parallel()
	input := []byte("the quick brown fox jumps over the lazy dog")
	first := HashDocument(input)
	second := HashDocument(input)
	if first != second {
		t.Fatalf("expected deterministic output, got %q and %q", first, second)
	}
}

func TestHashDocument_DifferentInputs(t *testing.T) {
	t.Parallel()
	a := HashDocument([]byte("input alpha"))
	b := HashDocument([]byte("input beta"))
	if a == b {
		t.Fatalf("different inputs produced identical hash: %q", a)
	}
}

func TestHashDocument_OutputFormat(t *testing.T) {
	t.Parallel()
	cases := []struct {
		name  string
		input []byte
	}{
		{"empty input", []byte{}},
		{"single byte", []byte{0x42}},
		{"short string", []byte("hello")},
		{"longer content", []byte(strings.Repeat("a", 4096))},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			hash := HashDocument(tc.input)
			if len(hash) != 64 {
				t.Fatalf("expected 64-char hex string, got %d chars: %q", len(hash), hash)
			}
			if _, err := hex.DecodeString(hash); err != nil {
				t.Fatalf("output is not valid hex: %q, err: %v", hash, err)
			}
		})
	}
}

func TestHashChunk_MatchesHashDocument(t *testing.T) {
	t.Parallel()
	input := []byte("chunk content for verification")
	docHash := HashDocument(input)
	chunkHash := HashChunk(input)
	if docHash != chunkHash {
		t.Fatalf("HashChunk and HashDocument diverge for same input: %q vs %q", chunkHash, docHash)
	}
}

func TestHashString_MatchesHashDocument(t *testing.T) {
	t.Parallel()
	s := "string input for verification"
	fromString := HashString(s)
	fromBytes := HashDocument([]byte(s))
	if fromString != fromBytes {
		t.Fatalf("HashString and HashDocument diverge: %q vs %q", fromString, fromBytes)
	}
}

func TestHashDocument_EmptyInput(t *testing.T) {
	t.Parallel()
	hash := HashDocument([]byte{})
	if len(hash) != 64 {
		t.Fatalf("expected 64-char hex for empty input, got %d chars", len(hash))
	}
	if _, err := hex.DecodeString(hash); err != nil {
		t.Fatalf("empty input hash is not valid hex: %v", err)
	}
}

func TestHashDocument_CrossSDKConformance(t *testing.T) {
	t.Parallel()
	// Known input/output pair for cross-SDK verification.
	// Input: "jeff's brain memory system"
	// Expected: BLAKE3 256-bit hex of that exact UTF-8 byte string.
	// The TS SDK test hardcodes the same expected value.
	input := []byte("jeff's brain memory system")
	hash := HashDocument(input)
	expected := "e311e54b56b26bfef4e5c8501f04c708f1e02233106022f58a7e94b728b7265c"
	if hash != expected {
		t.Fatalf("cross-SDK conformance failed:\n  got:  %s\n  want: %s", hash, expected)
	}
}

func TestHashSlug_Returns12HexChars(t *testing.T) {
	t.Parallel()
	cases := []struct {
		name  string
		input []byte
	}{
		{"empty", []byte{}},
		{"short", []byte("hello")},
		{"long", []byte(strings.Repeat("a", 4096))},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			slug := HashSlug(tc.input)
			if len(slug) != 12 {
				t.Fatalf("expected 12 chars, got %d: %q", len(slug), slug)
			}
			if _, err := hex.DecodeString(slug + "0000"); err != nil {
				// Validate it's valid hex by padding to even length check.
				for _, c := range slug {
					if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
						t.Fatalf("slug contains non-hex char: %q", slug)
					}
				}
			}
		})
	}
}

func TestHashSlug_IsPrefixOfHashDocument(t *testing.T) {
	t.Parallel()
	input := []byte("consistency check between HashSlug and HashDocument")
	full := HashDocument(input)
	slug := HashSlug(input)
	if full[:12] != slug {
		t.Fatalf("HashSlug is not a prefix of HashDocument: %q vs %q", slug, full[:12])
	}
}

func TestHashDocumentID_Returns16HexChars(t *testing.T) {
	t.Parallel()
	id := HashDocumentID("brain-1", "abc123")
	if len(id) != 16 {
		t.Fatalf("expected 16 chars, got %d: %q", len(id), id)
	}
	if _, err := hex.DecodeString(id); err != nil {
		t.Fatalf("output is not valid hex: %q, err: %v", id, err)
	}
}

func TestHashDocumentID_IsBrainScoped(t *testing.T) {
	t.Parallel()
	contentHash := "deadbeefcafebabe"
	idA := HashDocumentID("brain-alpha", contentHash)
	idB := HashDocumentID("brain-beta", contentHash)
	if idA == idB {
		t.Fatalf("different brains produced identical document IDs: %q", idA)
	}
}

func TestHashDocumentID_DifferentBrainsDifferentIDs(t *testing.T) {
	t.Parallel()
	contentHash := HashDocument([]byte("shared document content"))
	ids := make(map[string]string, 5)
	brains := []string{"brain-a", "brain-b", "brain-c", "brain-d", "brain-e"}
	for _, b := range brains {
		id := HashDocumentID(b, contentHash)
		if existing, ok := ids[id]; ok {
			t.Fatalf("collision: brain %q and %q both produced ID %q", existing, b, id)
		}
		ids[id] = b
	}
}

func TestHasher_BLAKE3HasherConformance(t *testing.T) {
	t.Parallel()
	var h Hasher = BLAKE3Hasher{}
	if h.Name() != "blake3" {
		t.Fatalf("expected name 'blake3', got %q", h.Name())
	}
	input := []byte("hasher interface test")
	got := h.Hash(input)
	want := HashDocument(input)
	if got != want {
		t.Fatalf("BLAKE3Hasher.Hash differs from HashDocument: %q vs %q", got, want)
	}
}

func BenchmarkBLAKE3_1KB(b *testing.B) {
	data := []byte(strings.Repeat("x", 1024))
	b.SetBytes(1024)
	b.ResetTimer()
	for b.Loop() {
		HashDocument(data)
	}
}

func BenchmarkBLAKE3_10KB(b *testing.B) {
	data := []byte(strings.Repeat("x", 10*1024))
	b.SetBytes(10 * 1024)
	b.ResetTimer()
	for b.Loop() {
		HashDocument(data)
	}
}

func BenchmarkBLAKE3_100KB(b *testing.B) {
	data := []byte(strings.Repeat("x", 100*1024))
	b.SetBytes(100 * 1024)
	b.ResetTimer()
	for b.Loop() {
		HashDocument(data)
	}
}

func BenchmarkSHA256_1KB(b *testing.B) {
	data := []byte(strings.Repeat("x", 1024))
	b.SetBytes(1024)
	b.ResetTimer()
	for b.Loop() {
		sum := sha256.Sum256(data)
		_ = hex.EncodeToString(sum[:])
	}
}

func BenchmarkSHA256_10KB(b *testing.B) {
	data := []byte(strings.Repeat("x", 10*1024))
	b.SetBytes(10 * 1024)
	b.ResetTimer()
	for b.Loop() {
		sum := sha256.Sum256(data)
		_ = hex.EncodeToString(sum[:])
	}
}

func BenchmarkSHA256_100KB(b *testing.B) {
	data := []byte(strings.Repeat("x", 100*1024))
	b.SetBytes(100 * 1024)
	b.ResetTimer()
	for b.Loop() {
		sum := sha256.Sum256(data)
		_ = hex.EncodeToString(sum[:])
	}
}
